test/AutoDiff/protocol_requirement_autodiff.swift

```diff
@@ -137,4 +137,60 @@ struct S : P {
   }
 }
 
+// MARK: - Overridden protocol method adding differentiable attribute.
+
+public protocol Distribution {
+  associatedtype Value
+  func logProbability(of value: Value) -> Float
+}
+
+public protocol DifferentiableDistribution: Differentiable, Distribution {
+  @differentiable(wrt: self)
+  func logProbability(of value: Value) -> Float
+}
+
+struct Foo: DifferentiableDistribution {
+  @differentiable(wrt: self)
+  func logProbability(of value: Float) -> Float {
+    .zero
+  }
+}
+
+@differentiable
+func blah<T: DifferentiableDistribution>(_ x: T) -> Float where T.Value: AdditiveArithmetic {
+  x.logProbability(of: .zero)
+}
+
+// Adding a more general `@differentiable` attribute.
+public protocol DoubleDifferentiableDistribution: DifferentiableDistribution
+  where Value: Differentiable {
+  @differentiable(wrt: self)
+  @differentiable(wrt: (self, value))
+  func logProbability(of value: Value) -> Float
+}
+
+@differentiable
+func blah2<T: DoubleDifferentiableDistribution>(_ x: T, _ value: T.Value) -> Float
+  where T.Value: AdditiveArithmetic {
+  x.logProbability(of: value)
+}
+
+// Missing a `@differentiable` attribute.
+public protocol MissingDifferentiableDistribution: DifferentiableDistribution
+  where Value: Differentiable {
+  func logProbability(of value: Value) -> Float // expected-note {{candidate is missing attribute '@differentiable(wrt: self)'}}
+}
+
+// Missing `@differentiable` attribute, without printing the 'wrt' arguments.
+protocol Example: Differentiable {
+  associatedtype Scalar: Differentiable
+
+  @differentiable
```

Comment:
Could you add a test where any `where` clause is rejected if specified in a `@differentiable` on a protocol requirement?

Reply:
This is not being checked for at the moment. Should I add a check as part of this PR? If so, where should this be done? As in, in which file?

Reply:
Should be done in `visitDifferentiableAttr` in TypeCheckAttr.cpp.

Reply:
Thanks. And how do I check if an `AbstractFunctionDecl` is declared inside a protocol or not?

Reply:
You can check whether getInnermostTypeContext() is a protocol decl.

Reply:
Done.

Reply:
Actually I think you should use `->isProtocolRequirement()` directly, as `getInnermostTypeContext()` would make `@differentiable` in protocol extensions fail unexpectedly.

Reply:
Done.